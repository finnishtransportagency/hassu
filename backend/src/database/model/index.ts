export * from "./common";
export * from "./projekti";
export * from "./suunnitteluVaihe";
export * from "./nahtavillaoloVaihe";
export * from "./hyvaksymisPaatosVaihe";
