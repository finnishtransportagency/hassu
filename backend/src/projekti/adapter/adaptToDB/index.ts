export * from "./adaptAloitusKuulutusToSave";
export * from "./adaptHyvaksymisPaatosVaiheToSave";
export * from "./adaptKayttajatunnusList";
export * from "./adaptNahtavillaoloVaiheToSave";
export * from "./adaptSuunnitteluSopimusToSave";
export * from "./adaptVuorovaikutusKierrosToSave";
export * from "./common";
