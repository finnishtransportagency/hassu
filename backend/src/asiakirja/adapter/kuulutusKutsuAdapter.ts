import { CommonKutsuAdapter, CommonKutsuAdapterProps } from "./commonKutsuAdapter";
import { IlmoituksenVastaanottajat, UudelleenKuulutus, Yhteystieto } from "../../database/model";
import { kuntametadata } from "hassu-common/kuntametadata";
import { formatNimi } from "../../util/userUtil";
import { organisaatioIsEly } from "../../util/organisaatioIsEly";
import { translate } from "../../util/localization";
import { formatDate } from "../asiakirjaUtil";
import { Kieli } from "hassu-common/graphql/apiModel";

export interface KuulutusKutsuAdapterProps extends CommonKutsuAdapterProps {
  kuulutusPaiva: string;
  kuulutusVaihePaattyyPaiva?: string;
  ilmoituksenVastaanottajat?: IlmoituksenVastaanottajat | null;
  uudelleenKuulutus?: UudelleenKuulutus | null;
  yhteystiedot: Yhteystieto[];
}

export abstract class KuulutusKutsuAdapter<T extends KuulutusKutsuAdapterProps> extends CommonKutsuAdapter {
  readonly ilmoituksenVastaanottajat: IlmoituksenVastaanottajat | null | undefined;
  readonly uudelleenKuulutus?: UudelleenKuulutus | null;
  readonly props: T;
  readonly vahainenMenettely: boolean | null | undefined;

  constructor(props: T, localizationKeyPrefix?: string) {
    super(props, localizationKeyPrefix);
    const { suunnitteluSopimus, ilmoituksenVastaanottajat, uudelleenKuulutus, vahainenMenettely } = props;
    this.suunnitteluSopimus = suunnitteluSopimus;
    this.ilmoituksenVastaanottajat = ilmoituksenVastaanottajat;
    this.uudelleenKuulutus = uudelleenKuulutus;
    this.vahainenMenettely = vahainenMenettely;
    this.props = props;
  }

  abstract get kuulutusNimiCapitalized(): string;
  abstract get kuulutusYllapitoUrl(): string;

  get kuulutusNimi(): string {
    return this.kuulutusNimiCapitalized.toLowerCase();
  }

  get laheteKirjeVastaanottajat(): string[] {
    const result: string[] = [];
    const kunnat = this.ilmoituksenVastaanottajat?.kunnat;
    const viranomaiset = this.ilmoituksenVastaanottajat?.viranomaiset;
    kunnat?.forEach(({ sahkoposti }) => {
      result.push(sahkoposti);
    });
    viranomaiset?.forEach(({ sahkoposti }) => {
      result.push(sahkoposti);
    });
    return result;
  }

  get uudelleenKuulutusSeloste(): string | undefined {
    return this.uudelleenKuulutus?.selosteLahetekirjeeseen?.[this.kieli];
  }

  get kuulutusPaiva(): string {
    return this.props.kuulutusPaiva ? formatDate(this.props.kuulutusPaiva) : "DD.MM.YYYY";
  }

  get kuulutusNahtavillaAika(): string {
    return this.formatDateRange(this.props.kuulutusPaiva, this.props.kuulutusVaihePaattyyPaiva);
  }

  get simple_yhteystiedot(): string[] {
    return this.props.yhteystiedot.map((y) => {
      let organisaatio = y.organisaatio;
      if (y.kunta) {
        organisaatio = kuntametadata.nameForKuntaId(y.kunta, this.kieli);
      } else if (organisaatioIsEly(y.organisaatio) && y.elyOrganisaatio) {
        const kaannos = translate(`viranomainen.${y.elyOrganisaatio}`, this.kieli);
        if (kaannos) {
          organisaatio = kaannos;
        }
      } else if (organisaatio) {
        const orgKey = organisaatio.toUpperCase().replace(/Ä/g, "A").replace(/Ö/g, "O").replace(/ /g, "_");
        const kaannos = translate(`viranomainen.${orgKey}`, this.kieli);
        if (kaannos) {
          organisaatio = kaannos;
        }
      }
      return `${organisaatio ? organisaatio + ", " : ""}${formatNimi(y)}, ${this.kieli === Kieli.RUOTSI ? "tel" : "puh"}. ${
        y.puhelinnumero
      }, ${y.sahkoposti}`;
    });
  }
}
