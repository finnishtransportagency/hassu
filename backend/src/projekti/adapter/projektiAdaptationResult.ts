import { DBProjekti } from "../../database/model";

export enum ProjektiEventType {
  VUOROVAIKUTUS_PUBLISHED = "VUOROVAIKUTUS_PUBLISHED",
  AINEISTO_CHANGED = "AINEISTO_CHANGED",
  LOGO_FILES_CHANGED = "LOGO_FILES_CHANGED",
  RESET_KAYTTOOIKEUDET = "RESET_KAYTTOOIKEUDET",
  SAVE_PROJEKTI_TO_VELHO = "SAVE_PROJEKTI_TO_VELHO",
  FILES_CHANGED = "FILES_CHANGED",
}

export type VuorovaikutusPublishedEvent = {
  eventType: ProjektiEventType.VUOROVAIKUTUS_PUBLISHED;
  vuorovaikutusNumero: number;
};

export type AineistoChangedEvent = { eventType: ProjektiEventType.AINEISTO_CHANGED };
export type logoFilesChangedEvent = { eventType: ProjektiEventType.LOGO_FILES_CHANGED };
export type ResetKayttooikeudetEvent = { eventType: ProjektiEventType.RESET_KAYTTOOIKEUDET };
export type SaveProjektiToVelhoEvent = { eventType: ProjektiEventType.SAVE_PROJEKTI_TO_VELHO };
export type FilesChangedEvent = { eventType: ProjektiEventType.FILES_CHANGED };

export type ProjektiEvent =
  | VuorovaikutusPublishedEvent
  | AineistoChangedEvent
  | ResetKayttooikeudetEvent
  | SaveProjektiToVelhoEvent
  | logoFilesChangedEvent
  | FilesChangedEvent;

export class ProjektiAdaptationResult {
  private dbProjekti: DBProjekti;
  private events: ProjektiEvent[] = [];

  constructor(dbProjekti: DBProjekti) {
    this.dbProjekti = dbProjekti;
  }

  setProjekti(dbProjekti: DBProjekti): void {
    this.dbProjekti = dbProjekti;
  }

  private pushEvent(event: ProjektiEvent): void {
    if (!this.events.find((e) => e.eventType == event.eventType)) {
      this.events.push(event);
    }
  }

  get projekti(): DBProjekti {
    return this.dbProjekti;
  }

  async onEvents(thisHasHappened: (events: ProjektiEvent[]) => boolean, eventHandler: (oid: string) => Promise<void>): Promise<void> {
    if (thisHasHappened(this.events)) {
      await eventHandler(this.projekti.oid);
    }
  }

  vuorovaikutusPublished(vuorovaikutusNumero: number): void {
    const newEvent: VuorovaikutusPublishedEvent = {
      eventType: ProjektiEventType.VUOROVAIKUTUS_PUBLISHED,
      vuorovaikutusNumero,
    };
    this.pushEvent(newEvent);
  }

  aineistoChanged(): void {
    const newEvent: AineistoChangedEvent = {
      eventType: ProjektiEventType.AINEISTO_CHANGED,
    };
    this.pushEvent(newEvent);
  }

  logoFilesChanged(): void {
    const newEvent: logoFilesChangedEvent = {
      eventType: ProjektiEventType.LOGO_FILES_CHANGED,
    };
    this.pushEvent(newEvent);
  }

  filesChanged(): void {
    const newEvent: FilesChangedEvent = {
      eventType: ProjektiEventType.FILES_CHANGED,
    };
    this.pushEvent(newEvent);
  }

  resetKayttooikeudetAndSynchronizeVelho(): void {
    const newEvent: ResetKayttooikeudetEvent = {
      eventType: ProjektiEventType.RESET_KAYTTOOIKEUDET,
    };
    this.pushEvent(newEvent);
  }

  saveProjektiToVelho(): void {
    const newEvent: SaveProjektiToVelhoEvent = {
      eventType: ProjektiEventType.SAVE_PROJEKTI_TO_VELHO,
    };
    this.pushEvent(newEvent);
  }
}
