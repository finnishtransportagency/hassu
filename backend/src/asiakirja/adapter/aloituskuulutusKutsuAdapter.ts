import { LokalisoituYhteystieto } from "./commonKutsuAdapter";
import {
  AloitusKuulutusJulkaisu,
  DBVaylaUser,
  IlmoituksenVastaanottajat,
  LocalizedMap,
  SuunnitteluSopimus,
  SuunnitteluSopimusJulkaisu,
  UudelleenKuulutus,
  Yhteystieto,
} from "../../database/model";
import { KayttajaTyyppi, KuulutusTekstit, LaskuriTyyppi } from "hassu-common/graphql/apiModel";
import { AsiakirjanMuoto } from "../asiakirjaTypes";
import { vaylaUserToYhteystieto, yhteystietoPlusKunta } from "../../util/vaylaUserToYhteystieto";
import { assertIsDefined } from "../../util/assertions";
import { kuntametadata } from "hassu-common/kuntametadata";
import { formatProperNoun } from "hassu-common/util/formatProperNoun";
import { calculateEndDate } from "../../endDateCalculator/endDateCalculatorHandler";
import { KaannettavaKieli } from "hassu-common/kaannettavatKielet";
import { KuulutusKutsuAdapter, KuulutusKutsuAdapterProps } from "./kuulutusKutsuAdapter";

type PropsCreatorOptions = {
  oid: string;
  lyhytOsoite: string | undefined | null;
  kayttoOikeudet: DBVaylaUser[];
  kieli: KaannettavaKieli;
  asianhallintaPaalla: boolean;
  linkkiAsianhallintaan: string | undefined;
  aloitusKuulutusJulkaisu: AloitusKuulutusJulkaisu | undefined;
  euRahoitusLogot: LocalizedMap<string> | null | undefined;
  vahainenMenettely: boolean | null | undefined;
  kuulutettuYhdessaSuunnitelmanimi: string | undefined;
};

export async function createAloituskuulutusKutsuAdapterProps({
  oid,
  lyhytOsoite,
  kayttoOikeudet,
  kieli,
  asianhallintaPaalla,
  linkkiAsianhallintaan,
  aloitusKuulutusJulkaisu,
  euRahoitusLogot,
  vahainenMenettely,
  kuulutettuYhdessaSuunnitelmanimi,
}: PropsCreatorOptions): Promise<AloituskuulutusKutsuAdapterProps> {
  assertIsDefined(aloitusKuulutusJulkaisu);
  assertIsDefined(aloitusKuulutusJulkaisu.kuulutusPaiva, "aloitusKuulutusJulkaisu.kuulutusPaiva puuttuu");
  const kuulutusVaihePaattyyPaiva = await calculateEndDate({
    alkupaiva: aloitusKuulutusJulkaisu.kuulutusPaiva,
    tyyppi: LaskuriTyyppi.KUULUTUKSEN_PAATTYMISPAIVA,
  });
  return {
    oid,
    lyhytOsoite,
    hankkeenKuvaus: aloitusKuulutusJulkaisu.hankkeenKuvaus,
    kieli,
    kielitiedot: aloitusKuulutusJulkaisu.kielitiedot,
    kuulutusPaiva: aloitusKuulutusJulkaisu.kuulutusPaiva,
    kuulutusVaihePaattyyPaiva,
    velho: aloitusKuulutusJulkaisu.velho,
    yhteystiedot: aloitusKuulutusJulkaisu.yhteystiedot,
    suunnitteluSopimus: aloitusKuulutusJulkaisu.suunnitteluSopimus ?? undefined,
    kayttoOikeudet,
    uudelleenKuulutus: aloitusKuulutusJulkaisu.uudelleenKuulutus ?? undefined,
    euRahoitusLogot: euRahoitusLogot ?? undefined,
    vahainenMenettely,
    asianhallintaPaalla,
    linkkiAsianhallintaan,
    kuulutettuYhdessaSuunnitelmanimi,
  };
}

export interface AloituskuulutusKutsuAdapterProps extends KuulutusKutsuAdapterProps {}

export class AloituskuulutusKutsuAdapter extends KuulutusKutsuAdapter<AloituskuulutusKutsuAdapterProps> {
  readonly suunnitteluSopimus?: SuunnitteluSopimusJulkaisu | SuunnitteluSopimus | null;
  readonly ilmoituksenVastaanottajat: IlmoituksenVastaanottajat | null | undefined;
  readonly uudelleenKuulutus?: UudelleenKuulutus | null;
  readonly vahainenMenettely: boolean | null | undefined;

  constructor(props: AloituskuulutusKutsuAdapterProps) {
    super(props);
    const { suunnitteluSopimus, ilmoituksenVastaanottajat, uudelleenKuulutus, vahainenMenettely } = props;
    this.suunnitteluSopimus = suunnitteluSopimus;
    this.ilmoituksenVastaanottajat = ilmoituksenVastaanottajat;
    this.uudelleenKuulutus = uudelleenKuulutus;
    this.vahainenMenettely = vahainenMenettely;
  }

  get kuulutusNimiCapitalized(): string {
    return "Aloituskuulutus";
  }

  get kuulutusYllapitoUrl(): string {
    return super.aloituskuulutusYllapitoUrl;
  }

  yhteystiedot(
    yhteystiedot: Yhteystieto[] | null | undefined,
    yhteysHenkilot?: string[] | null,
    pakotaProjariTaiKunnanEdustaja?: boolean
  ): LokalisoituYhteystieto[] {
    let yt: Yhteystieto[] = [];
    let suunnitteluSopimus: SuunnitteluSopimusJulkaisu;
    const kunnanEdustaja = this.kayttoOikeudet?.find(
      (ko) =>
        ko.email === (this.suunnitteluSopimus as SuunnitteluSopimusJulkaisu)?.email ||
        ko.kayttajatunnus === (this.suunnitteluSopimus as SuunnitteluSopimus)?.yhteysHenkilo
    );
    if (kunnanEdustaja && this.suunnitteluSopimus) {
      suunnitteluSopimus = {
        ...kunnanEdustaja,
        kunta: this.suunnitteluSopimus.kunta,
        puhelinnumero: kunnanEdustaja.puhelinnumero ?? "",
      };
    }
    if (yhteystiedot) {
      yt = yt.concat(yhteystiedot.map((yt) => yhteystietoPlusKunta(yt, suunnitteluSopimus)));
    }
    if (yhteysHenkilot) {
      if (!this.kayttoOikeudet) {
        throw new Error("BUG: Kayttöoikeudet pitää antaa jos yhteyshenkilöt on annettu.");
      }
      this.getUsersForUsernames(yhteysHenkilot).forEach((user) => {
        yt.push(vaylaUserToYhteystieto(user, this.suunnitteluSopimus));
      });
    }
    if (pakotaProjariTaiKunnanEdustaja) {
      const projari = this.kayttoOikeudet?.find((ko) => ko.tyyppi == KayttajaTyyppi.PROJEKTIPAALLIKKO);

      if (this.suunnitteluSopimus && !yt.find((t) => t.sahkoposti === kunnanEdustaja?.email)) {
        yt = [vaylaUserToYhteystieto(kunnanEdustaja as DBVaylaUser, this.suunnitteluSopimus)].concat(yt);
      } else if (!yt.find((t) => t.sahkoposti === projari?.email)) {
        yt = [vaylaUserToYhteystieto(projari as DBVaylaUser, this.suunnitteluSopimus)].concat(yt);
      }
    }

    return yt.map((yt) => this.yhteystietoMapper(yt));
  }

  get suunnittelusopimusKunta(): string {
    const suunnitteluSopimus = this.suunnitteluSopimus;
    if (this.isOsapuoletInUse) {
      const nimet = this.props
        .suunnitteluSopimus!.osapuolet!.map((osapuoli) => osapuoli.osapuolenNimiEnsisijainen)
        .filter((nimi) => nimi !== undefined && nimi !== null);
      if (nimet.length > 0) {
        return nimet.join(", ");
      }
    }
    assertIsDefined(suunnitteluSopimus?.kunta, "Suunnittelusopimuksella pitää olla kunta");
    return kuntametadata.nameForKuntaId(suunnitteluSopimus?.kunta, this.kieli);
  }

  get suunnitelmasta_vastaavalla_tai_rataverkon_haltijalla(): string {
    if (this.asiakirjanMuoto == AsiakirjanMuoto.RATA) {
      return this.text("asiakirja.aloituskuulutus.rataverkon_haltijalla");
    }
    return this.text("asiakirja.aloituskuulutus.suunnitelmasta_vastaavalla");
  }

  get lakiviite_tutkimusoikeus(): string {
    if (this.asiakirjanMuoto == AsiakirjanMuoto.RATA) {
      return this.text("asiakirja.aloituskuulutus.lakiviite_tutkimusoikeus_rata");
    }
    return this.text("asiakirja.aloituskuulutus.lakiviite_tutkimusoikeus_tie");
  }

  get lakiviite_muistutus(): string {
    if (this.asiakirjanMuoto == AsiakirjanMuoto.RATA) {
      return this.text("asiakirja.aloituskuulutus.lakiviite_muistutus_rata");
    }
    return this.text("asiakirja.aloituskuulutus.lakiviite_muistutus_tie");
  }

  get lakiviite_vahainen_menettely(): string {
    if (this.asiakirjanMuoto == AsiakirjanMuoto.RATA) {
      return this.text("asiakirja.aloituskuulutus.lakiviite_vahainen_menettely_rata");
    }
    return this.text("asiakirja.aloituskuulutus.lakiviite_vahainen_menettely_tie");
  }

  get lakiviite_kappale4_vahainen_menettely(): string {
    if (this.asiakirjanMuoto == AsiakirjanMuoto.RATA) {
      return this.text("asiakirja.aloituskuulutus.lakiviite_kappale4_vahainen_menettely_rata");
    }
    return this.text("asiakirja.aloituskuulutus.lakiviite_kappale4_vahainen_menettely_tie");
  }

  get aloituskuulutus_vahainen_menettely_lakiviite(): string {
    if (this.asiakirjanMuoto == AsiakirjanMuoto.RATA) {
      return this.text("asiakirja.aloituskuulutus_vahainen_menettely_lakiviite_rata");
    }
    return this.text("asiakirja.aloituskuulutus_vahainen_menettely_lakiviite_tie");
  }

  get kuuluttaja(): string {
    const suunnitteluSopimus = this.suunnitteluSopimus;
    if (suunnitteluSopimus?.kunta) {
      return formatProperNoun(kuntametadata.nameForKuntaId(suunnitteluSopimus.kunta, this.kieli));
    }
    return super.kuuluttaja;
  }

  get kuuluttaja_pitka(): string {
    const suunnitteluSopimus = this.suunnitteluSopimus;
    if (suunnitteluSopimus?.kunta) {
      return formatProperNoun(kuntametadata.nameForKuntaId(suunnitteluSopimus.kunta, this.kieli));
    }
    return super.kuuluttaja_pitka;
  }

  get isOsapuoletInUse(): boolean {
    if (this.props.suunnitteluSopimus && this.props.suunnitteluSopimus.osapuolet && this.props.suunnitteluSopimus.osapuolet.length > 0) {
      return true;
    }
    return false;
  }

  kutsuja(): string | undefined {
    if (this.props.suunnitteluSopimus && this.props.suunnitteluSopimus.kunta && this.props.suunnitteluSopimus.kunta !== undefined) {
      return kuntametadata.nameForKuntaId(this.props.suunnitteluSopimus.kunta, this.kieli);
    }
    if (this.isOsapuoletInUse) {
      const nimet = this.props
        .suunnitteluSopimus!.osapuolet!.map((osapuoli) => osapuoli.osapuolenNimiEnsisijainen)
        .filter((nimi) => nimi !== undefined && nimi !== null);
      if (nimet.length > 0) {
        return nimet.join(", ");
      }
    }
    return super.kutsuja();
  }

  get userInterfaceFields(): KuulutusTekstit {
    let kappale1;
    if (this.suunnitteluSopimus) {
      kappale1 = this.htmlText("asiakirja.aloituskuulutus.kappale1_suunnittelusopimus");
    } else if (this.vahainenMenettely) {
      kappale1 = this.htmlText("asiakirja.aloituskuulutus.kappale1_vahainen_menettely");
    } else {
      kappale1 = this.htmlText("asiakirja.aloituskuulutus.kappale1");
    }
    return {
      __typename: "KuulutusTekstit",
      leipaTekstit: [kappale1],
      kuvausTekstit: [this.htmlText("asiakirja.aloituskuulutus.kappale2_ui")],
      infoTekstit: [
        this.htmlText("asiakirja.aloituskuulutus.kappale3"),
        this.vahainenMenettely
          ? this.htmlText("asiakirja.aloituskuulutus.kappale4_vahainen_menettely")
          : this.htmlText("asiakirja.aloituskuulutus.kappale4"),
      ],
      tietosuoja: this.htmlText("asiakirja.tietosuoja", { extLinks: true }),
    };
  }
}
