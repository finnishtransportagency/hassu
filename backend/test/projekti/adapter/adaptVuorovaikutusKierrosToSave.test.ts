import { describe, it } from "mocha";
import { adaptVuorovaikutusKierrosToSave } from "../../../src/projekti/adapter/adaptToDB";
import { ProjektiAdaptationResult } from "../../../src/projekti/adapter/projektiAdaptationResult";
import { ProjektiFixture } from "../../fixture/projektiFixture";
import {
  IlmoitettavaViranomainen,
  KaytettavaPalvelu,
  VuorovaikutusKierrosInput,
  VuorovaikutusKierrosTila,
  VuorovaikutusTilaisuusTyyppi,
} from "hassu-common/graphql/apiModel";
import { VuorovaikutusKierros } from "../../../src/database/model";
import { cloneDeep } from "lodash";

import { expect } from "chai";

describe("adaptVuorovaikutusKierrosToSave", () => {
  it("should adapt vuorovaikutuskierros to save", async () => {
    const dbProjekti = new ProjektiFixture().dbProjekti1();
    delete dbProjekti.kielitiedot?.toissijainenKieli;
    const vuorovaikutusKierrosInput = {
      ilmoituksenVastaanottajat: {
        viranomaiset: [
          {
            nimi: IlmoitettavaViranomainen.KAINUUN_ELY,
            sahkoposti: "kirjaamo.kainuu@ely-keskus.fi",
          },
        ],
        kunnat: [
          {
            id: 620,
            sahkoposti: "maikol@iki.fi",
          },
        ],
      },
      vuorovaikutusTilaisuudet: [
        {
          nimi: {
            SUOMI: "fdsaadfs",
          },
          paivamaara: "2023-02-23",
          alkamisAika: "00:00",
          paattymisAika: "00:00",
          tyyppi: VuorovaikutusTilaisuusTyyppi.PAIKALLA,
          paikka: {
            SUOMI: "adfsdfas",
          },
          osoite: {
            SUOMI: "dfsasdaf",
          },
          postinumero: "12344",
          postitoimipaikka: {
            SUOMI: "",
          },
          lisatiedot: {
            SUOMI: "",
          },
        },
        {
          nimi: {
            SUOMI: "fdsaadfs",
          },
          paivamaara: "2023-02-23",
          alkamisAika: "00:00",
          paattymisAika: "00:00",
          tyyppi: VuorovaikutusTilaisuusTyyppi.VERKOSSA,
          kaytettavaPalvelu: KaytettavaPalvelu.TEAMS,
          linkki: "http://www.fi",
          lisatiedot: {
            SUOMI: "Voit liittyä tilaisuuteen myös puhelimella soittamalla numeroon 05012345.",
          },
        },
      ],
      esitettavatYhteystiedot: {
        yhteysTiedot: [],
        yhteysHenkilot: [],
      },
      vuorovaikutusJulkaisuPaiva: "2023-02-20",
      hankkeenKuvaus: {
        SUOMI: "Kuvaus",
      },
      vuorovaikutusNumero: 1,
      kysymyksetJaPalautteetViimeistaan: "2023-02-22",
    };

    const result = adaptVuorovaikutusKierrosToSave(dbProjekti, vuorovaikutusKierrosInput, new ProjektiAdaptationResult(dbProjekti));

    expect(result).toMatchSnapshot();
  });

  it("should not null keys that are not given", async () => {
    const dbProjekti = new ProjektiFixture().dbProjektiLackingNahtavillaoloVaihe();
    dbProjekti.vuorovaikutusKierros = {
      ...dbProjekti.vuorovaikutusKierros,
      vuorovaikutusNumero: 1,
      tila: VuorovaikutusKierrosTila.MUOKATTAVISSA,
    };
    const vuorovaikutusKierrosInput: VuorovaikutusKierrosInput = {
      vuorovaikutusNumero: 1,
      esittelyaineistot: [],
    };

    const expectedResult: VuorovaikutusKierros = cloneDeep({
      ...dbProjekti.vuorovaikutusKierros,
      aineistot: [],
    });

    const result = adaptVuorovaikutusKierrosToSave(dbProjekti, vuorovaikutusKierrosInput, new ProjektiAdaptationResult(dbProjekti));

    expect(result).to.eql(expectedResult);
  });

  it("should allow esitettavatYhteystiedot.yhteystiedot to be emptied", async () => {
    const dbProjekti = new ProjektiFixture().dbProjektiLackingNahtavillaoloVaihe();
    const vuorovaikutusNumero = 1;
    dbProjekti.vuorovaikutusKierros = {
      ...dbProjekti.vuorovaikutusKierros,
      esitettavatYhteystiedot: {
        yhteysTiedot: [
          {
            etunimi: "etunimi",
            puhelinnumero: "123456",
            sahkoposti: "sahkoposti@sahkoposti.fi",
            sukunimi: "sukunimi",
            elyOrganisaatio: undefined,
            kunta: undefined,
            organisaatio: "organisaatio",
            titteli: undefined,
          },
        ],
        yhteysHenkilot: ["A000111"],
      },
      vuorovaikutusNumero,
      tila: VuorovaikutusKierrosTila.MUOKATTAVISSA,
    };

    const result = adaptVuorovaikutusKierrosToSave(
      dbProjekti,
      { vuorovaikutusNumero, esitettavatYhteystiedot: { yhteysHenkilot: [], yhteysTiedot: [] } },
      new ProjektiAdaptationResult(dbProjekti)
    );

    expect(result?.esitettavatYhteystiedot?.yhteysTiedot?.length).to.eql(0);
  });
});
