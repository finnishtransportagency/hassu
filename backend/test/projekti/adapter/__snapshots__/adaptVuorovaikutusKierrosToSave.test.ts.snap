// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adaptVuorovaikutusKierrosToSave should adapt vuorovaikutuskierros to save 1`] = `
Object {
  "esitettavatYhteystiedot": Object {
    "yhteysHenkilot": Array [],
    "yhteysTiedot": Array [],
  },
  "hankkeenKuvaus": Object {
    "SUOMI": "Kuvaus",
  },
  "ilmoituksenVastaanottajat": Object {
    "kunnat": Array [
      Object {
        "id": 620,
        "sahkoposti": "maikol@iki.fi",
      },
    ],
    "viranomaiset": Array [
      Object {
        "nimi": "KAINUUN_ELY",
        "sahkoposti": "kirjaamo.kainuu@ely-keskus.fi",
      },
    ],
  },
  "kysymyksetJaPalautteetViimeistaan": "2023-02-22",
  "tila": "MUOKATTAVISSA",
  "vuorovaikutusJulkaisuPaiva": "2023-02-20",
  "vuorovaikutusNumero": 1,
  "vuorovaikutusTilaisuudet": Array [
    Object {
      "alkamisAika": "00:00",
      "lisatiedot": Object {
        "SUOMI": "",
      },
      "nimi": Object {
        "SUOMI": "fdsaadfs",
      },
      "osoite": Object {
        "SUOMI": "dfsasdaf",
      },
      "paattymisAika": "00:00",
      "paikka": Object {
        "SUOMI": "adfsdfas",
      },
      "paivamaara": "2023-02-23",
      "postinumero": "12344",
      "postitoimipaikka": Object {
        "SUOMI": "",
      },
      "tyyppi": "PAIKALLA",
    },
    Object {
      "alkamisAika": "00:00",
      "kaytettavaPalvelu": "TEAMS",
      "linkki": "http://www.fi",
      "lisatiedot": Object {
        "SUOMI": "Voit liittyä tilaisuuteen myös puhelimella soittamalla numeroon 05012345.",
      },
      "nimi": Object {
        "SUOMI": "fdsaadfs",
      },
      "paattymisAika": "00:00",
      "paivamaara": "2023-02-23",
      "tyyppi": "VERKOSSA",
    },
  ],
}
`;
